File: leetcode/652-寻找重复的子树/652-寻找重复的子树.py
# Definition for a binary tree node.
from re import sub
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 652-寻找重复的子树
class Solution:
    
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        memo = {}
        res = []
        def traverse(root:Optional[TreeNode]):
            if root is None:
                return "#"
            left = traverse(root.left)
            right = traverse(root.right)
            sub_tree = "{} - {} - {}".format(left,right,root.val)
            if sub_tree in memo.keys():
                memo[sub_tree] = memo[sub_tree]+1
            else:
                memo[sub_tree] = 1
            if memo.get(sub_tree,0) == 2:
                res.append(root)
            
            return sub_tree
        traverse(root)
        return res

File: leetcode/652-寻找重复的子树/652-寻找重复的子树.js
/**
 * Definition for a binary tree node.
*/
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {TreeNode[]}
 */

// 652-寻找重复的子树
var findDuplicateSubtrees = function (root) {
    let memo = new Map();
    let res = new Array();
    const traverse = function (root) {
        if (!root) {
            return "#";
        }
        const left = traverse(root.left);
        const right = traverse(root.right);
        const subTree = `${left},${right},${root.val}`;
        const count = memo.get(subTree);
        memo.set(subTree, (count || 0) + 1);
        if (memo.get(subTree) == 2) res.push(root);
        return subTree;
    };
    traverse(root);
    return res;
};